File: frontend/services/studentService.ts
import apiClient from '@/lib/api-client';
import {
  Student,
  CreateStudentRequest,
  UpdateStudentRequest,
  StudentSearchParams,
  PageResponse,
  StudentListItem,
} from '@/types';

export const studentService = {
  /**
   * Create a new student
   */
  async createStudent(data: CreateStudentRequest): Promise<Student> {
    const response = await apiClient.post<Student>('/students', data);
    return response.data;
  },

  /**
   * Get student by key
   */
  async getStudent(studentKey: string): Promise<Student> {
    const response = await apiClient.get<Student>(`/students/${studentKey}`);
    return response.data;
  },

  /**
   * Search students with pagination
   */
  async searchStudents(params: StudentSearchParams = {}): Promise<PageResponse<StudentListItem>> {
    const queryParams: Record<string, any> = {
      page: params.page ?? 0,
      size: params.size ?? 20,
    };

    if (params.lastName) {
      queryParams.lastName = params.lastName;
    }

    if (params.guardianName) {
      queryParams.guardianName = params.guardianName;
    }

    if (params.status) {
      queryParams.status = params.status;
    }

    if (params.sort) {
      queryParams.sort = params.sort;
    }

    const response = await apiClient.get<PageResponse<StudentListItem>>('/students', {
      params: queryParams,
    });

    return response.data;
  },

  /**
   * Update student
   */
  async updateStudent(studentKey: string, data: UpdateStudentRequest): Promise<Student> {
    const response = await apiClient.put<Student>(`/students/${studentKey}`, data);
    return response.data;
  },

  /**
   * Delete student
   */
  async deleteStudent(studentKey: string): Promise<void> {
    await apiClient.delete(`/students/${studentKey}`);
  },

  /**
   * Get all students (without pagination) - for exports or bulk operations
   */
  async getAllStudents(): Promise<Student[]> {
    const response = await apiClient.get<PageResponse<StudentListItem>>('/students', {
      params: { size: 1000 },
    });
    return response.data.content as any;
  },
};

File: backend/student-service/src/main/java/com/sms/student/application/service/StudentService.java
package com.sms.student.application.service;

import com.sms.student.presentation.dto.StudentDTO;
import com.sms.student.presentation.dto.CreateStudentRequest;
import org.springframework.data.domain.Page;

/**
 * Service interface for student management business logic.
 * Defines the contract for student operations.
 */
public interface StudentService {

    /**
     * Create a new student.
     * Generates student key, validates age (3-18), checks mobile uniqueness.
     *
     * @param request the request containing student details
     * @return the created student DTO
     * @throws com.sms.shared.exception.ValidationException if age is invalid
     * @throws com.sms.shared.exception.DuplicateResourceException if mobile already exists
     */
    StudentDTO createStudent(CreateStudentRequest request);

    /**
     * Get student by student key.
     *
     * @param studentKey the student business key
     * @return the student DTO
     * @throws com.sms.shared.exception.ResourceNotFoundException if student not found
     */
    StudentDTO getStudent(String studentKey);

    /**
     * Update student by student key.
     * Only allows updating firstName, lastName, mobile, and status.
     *
     * @param studentKey the student business key
     * @param request the update request
     * @return the updated student DTO
     * @throws com.sms.shared.exception.ResourceNotFoundException if student not found
     * @throws com.sms.shared.exception.DuplicateResourceException if mobile is already taken
     */
    StudentDTO updateStudent(String studentKey, CreateStudentRequest request);

    /**
     * Delete a student by student key.
     *
     * @param studentKey the student business key
     * @throws com.sms.shared.exception.ResourceNotFoundException if student not found
     */
    void deleteStudent(String studentKey);

    /**
     * Search students by lastName or fatherNameOrGuardian.
     *
     * @param searchTerm the search term
     * @param page the page number (0-indexed)
     * @param size the page size
     * @return page of students matching the search term
     */
    Page<StudentDTO> searchStudents(String searchTerm, int page, int size);
}
